File: benchmarks/python/data_bench.py
"""Python data benchmarks — using pandas/polars for comparison."""
import time
import csv
import tempfile
import os

def create_test_csv(rows):
    f = tempfile.NamedTemporaryFile(mode='w', suffix='.csv', delete=False)
    writer = csv.writer(f)
    writer.writerow(['id', 'name', 'value', 'category'])
    for i in range(rows):
        cat = 'A' if i % 3 == 0 else ('B' if i % 3 == 1 else 'C')
        writer.writerow([i, f'item_{i}', i * 1.5, cat])
    f.close()
    return f.name

def bench_pandas_pipeline(csv_path):
    try:
        import pandas as pd
    except ImportError:
        print("pandas not installed, skipping")
        return

    start = time.perf_counter()
    iterations = 5
    for _ in range(iterations):
        df = pd.read_csv(csv_path)
        result = (df[df['value'] > 100.0]
                  [['id', 'name', 'value']]
                  .sort_values('value'))
        count = len(result)
    elapsed = (time.perf_counter() - start) / iterations
    print(f"100k pipeline pandas:   {elapsed*1000:.2f} ms  (rows={count})")

def bench_polars_pipeline(csv_path):
    try:
        import polars as pl
    except ImportError:
        print("polars not installed, skipping")
        return

    start = time.perf_counter()
    iterations = 5
    for _ in range(iterations):
        df = pl.read_csv(csv_path)
        result = (df.filter(pl.col('value') > 100.0)
                  .select(['id', 'name', 'value'])
                  .sort('value'))
        count = len(result)
    elapsed = (time.perf_counter() - start) / iterations
    print(f"100k pipeline polars:   {elapsed*1000:.2f} ms  (rows={count})")

if __name__ == "__main__":
    csv_path = create_test_csv(100_000)
    try:
        bench_pandas_pipeline(csv_path)
        bench_polars_pipeline(csv_path)
    finally:
        os.unlink(csv_path)

File: benchmarks/benches/data_bench.rs
// Data benchmarks: CSV pipeline with DataFusion.

use criterion::{Criterion, criterion_group, criterion_main};
use std::io::Write;
use tempfile::NamedTempFile;
use tl_compiler::{Vm, compile};
use tl_interpreter::Interpreter;
use tl_parser::parse;

fn create_test_csv(rows: usize) -> NamedTempFile {
    let mut f = NamedTempFile::new().unwrap();
    writeln!(f, "id,name,value,category").unwrap();
    for i in 0..rows {
        let cat = if i % 3 == 0 {
            "A"
        } else if i % 3 == 1 {
            "B"
        } else {
            "C"
        };
        writeln!(f, "{},item_{},{},{}", i, i, (i as f64) * 1.5, cat).unwrap();
    }
    f.flush().unwrap();
    f
}

fn data_pipeline_source(csv_path: &str) -> String {
    format!(
        r#"
let data = read_csv("{csv_path}")
let result = data
    |> filter(value > 100.0)
    |> select(id, name, value)
    |> sort(value)
let batches = collect(result)
len(batches)
"#,
    )
}

fn data_aggregate_source(csv_path: &str) -> String {
    format!(
        r#"
let data = read_csv("{csv_path}")
let result = data
    |> aggregate(avg(value), count(id))
collect(result)
"#,
    )
}

fn bench_data_pipeline_interpreter(c: &mut Criterion) {
    let csv = create_test_csv(100_000);
    let path = csv.path().to_str().unwrap().replace('\\', "/");
    let source = data_pipeline_source(&path);
    let program = parse(&source).unwrap();

    c.bench_function("100k row pipeline interpreter", |b| {
        b.iter(|| {
            let mut interp = Interpreter::new();
            interp.execute(&program).unwrap();
        })
    });
}

fn bench_data_pipeline_vm(c: &mut Criterion) {
    let csv = create_test_csv(100_000);
    let path = csv.path().to_str().unwrap().replace('\\', "/");
    let source = data_pipeline_source(&path);
    let program = parse(&source).unwrap();
    let proto = compile(&program).unwrap();

    c.bench_function("100k row pipeline VM", |b| {
        b.iter(|| {
            let mut vm = Vm::new();
            vm.execute(&proto).unwrap();
        })
    });
}

fn bench_data_aggregate_interpreter(c: &mut Criterion) {
    let csv = create_test_csv(100_000);
    let path = csv.path().to_str().unwrap().replace('\\', "/");
    let source = data_aggregate_source(&path);
    let program = parse(&source).unwrap();

    c.bench_function("100k aggregate interpreter", |b| {
        b.iter(|| {
            let mut interp = Interpreter::new();
            interp.execute(&program).unwrap();
        })
    });
}

fn bench_data_aggregate_vm(c: &mut Criterion) {
    let csv = create_test_csv(100_000);
    let path = csv.path().to_str().unwrap().replace('\\', "/");
    let source = data_aggregate_source(&path);
    let program = parse(&source).unwrap();
    let proto = compile(&program).unwrap();

    c.bench_function("100k aggregate VM", |b| {
        b.iter(|| {
            let mut vm = Vm::new();
            vm.execute(&proto).unwrap();
        })
    });
}

criterion_group!(
    benches,
    bench_data_pipeline_interpreter,
    bench_data_pipeline_vm,
    bench_data_aggregate_interpreter,
    bench_data_aggregate_vm,
);
criterion_main!(benches);
